File: _assets/scripts/normalizers/particles_normalize_old.lua
return function(data)
    local data_attr = python.as_attrgetter(data)
    if not data_attr.__contains__("particles") then
        return
    end
    local particle_identifier = nil
    for index, key in python.enumerate(data["particles"]) do
        particle_identifier = key
    end
    assert(particle_identifier ~= nil, "dictionary is empty; cannot find particle identifier")
    local output = python.builtins.dict()
    output["format_version"] = data["format_version"]
    output["defined_in"] = data["defined_in"]
    output["particle_effect"] = data["particles"][particle_identifier]
    output["particle_effect"]["description"] = python.builtins.dict()
    output["particle_effect"]["description"]["basic_render_parameters"] = data["particles"][particle_identifier]["basic_render_parameters"]
    output["particle_effect"]["description"]["identifier"] = particle_identifier
    python.as_attrgetter(output["particle_effect"]).__delitem__("basic_render_parameters")
    return output
end

File: _assets/scripts/normalizers/particles_normalize_old.py
from typing import Any, TypedDict, cast

__all__ = ["particles_normalize_old"]

description_typed_dict = TypedDict("description_typed_dict", {
    "basic_render_parameters": dict[str,Any],
    "identifier": str,
})

particle_effect_typed_dict = TypedDict("particle_effect_typed_dict", {
    "description": description_typed_dict,
})

output_type = TypedDict("output_type", {
    "format_version": str,
    "particle_effect": particle_effect_typed_dict,
})

input_type = dict[str,Any]

def particles_normalize_old(data:input_type|output_type) -> output_type|None:
    if "particles" not in data:
        return
    particle_identifier = list(data["particles"].keys())[0]
    output:output_type = {
        "format_version": cast(str, data["format_version"]),
        "particle_effect": data["particles"][particle_identifier],
    }
    output["particle_effect"]["description"] = {
        "basic_render_parameters": data["particles"][particle_identifier]["basic_render_parameters"],
        "identifier": particle_identifier,
    }
    del output["particle_effect"]["basic_render_parameters"] # type: ignore
    return output
